controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/ArtifactExpirer.java

```diff
@@ -73,4 +92,35 @@ private boolean isExpired(Artifact artifact, Instant now, VersionStatus versionS
         return true;
     }
 
+    /** Model versions in use in this system, and, if this is a CD system, in the main/public system */
+    private Set<Version> modelVersionsInUse() {
+        var system = controller().system();
+        var versions = versionsForSystem(system);
+
+        if (controller().system().isCd()) {
+            if (system == SystemName.PublicCd)
+                versions.addAll(versionsForSystem(SystemName.Public));
+            else if (system == SystemName.cd)
+                versions.addAll(versionsForSystem(SystemName.main));
+        }
+
+        log.log(INFO, "model versions in use : " + versions);
+        return versions;
+    }
+
+    private Set<Version> versionsForSystem(SystemName systemName) {
+        var versions = readConfigModelVersionsForSystem(systemName.name());
+        log.log(INFO, "versions for system " + systemName.name() + ": " + versions);
+        return versions;
+    }
+
+    private Set<Version> readConfigModelVersionsForSystem(String systemName) {
+        List<String> lines = uncheck(() -> Files.readAllLines(configModelPath.resolve("config-models-" + systemName + ".xml")));
+        return lines.stream()
+                .filter(line -> line.contains("VespaModelFactory."))
+                .map(line -> line.substring(line.indexOf("id='VespaModelFactory") + 22, line.indexOf("' class")))
```

Comment:
`+ 22` comes from `"id='VespaModelFactory".size()` ?  Extract `"id='VespaModelFactory"` as a constant?

With `startIndex = line.indexOf("id='VespaModelFactory") + 22`, the end index of the version would be `lineIndex.indexOf("'", startIndex)`, which avoids relying on the presence of the following `class` tag.